016-kafka-connect-build.md

```diff
@@ -0,0 +1,127 @@
+This does not have to run within the same Kubernetes cluster - it can be also SaaS registry such as Docker Hub or Quay.io.
+
+Following is an example of the Kafka Connect CR:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta1
+kind: KafkaConnect
+metadata:
+  name: my-connect-cluster
+  annotations:
+    strimzi.io/use-connector-resources: "true"
+spec:
+  version: 2.6.0
+  replicas: 1
+  bootstrapServers: my-cluster-kafka-bootstrap:9093
+  build:
+    output:
+      type: docker
+      image: docker.io/user/image:tag
+      pushSecret: dockerhub-credentaials
+    plugins:
+      - name: my-connector
+        artifacts:
+          - type: jar
+            url: https://some.url/artifact.jar
+            sha512sum: abc123
+          - type: jar
+            url: https://some.url/artifact2.jar
+            sha512sum: def456
+      - name: my-connector2
+        # ...
+  tls:
+    trustedCertificates:
+      - secretName: my-cluster-cluster-ca-cert
+        certificate: ca.crt
+  config:
+    group.id: connect-cluster
+    offset.storage.topic: connect-cluster-offsets
+    config.storage.topic: connect-cluster-configs
+    status.storage.topic: connect-cluster-status
+```
+
+When the `build` section is specified in the `KafkaConnect` resource, Strimzi will generate `Dockerfile` which will correspond to the configured connectors and their artifacts.
+It will build the container image from this `Dockerfile` in a separate pod and push it to the configured output (Docker registry / ImageStream).
+It will take the digest of this image and configure the Kafka Connect cluster to use this image.
+
+The build will use two different technologies:
+* [Kaniko](https://github.com/GoogleContainerTools/kaniko) will be used on Kubernetes
+* OpenShift Builds will be used on OKD / OpenShift 
+
+Kaniko is a daemon-less container builder which does not require any special privileges and runs in user-space.
+OpenShift Builds are part of the OKD / OpenShift Kubernetes platform and also do not require any special privileges.
+So it should work for most users across wide variate of environments.
+
+## Affected/not affected projects
+
+The existing `KafkaConnectS2I` resource is not affected by it.
+But it should be deprecated and later removed.
+
+`KafkaMirrorMaker2` is also based on Kafka Connect.
+But the build will not be available for `KafkaMirrorMaker2`.
+
+## Compatibility
+
+This proposal is fully backwards-compatible.
+When users don't specify the `build` section inside the `KafkaConnect` CR, the build will be never activated and used.
+So existing users who do not the new build will not be affected in any way.
+They will be also still able to use the old manually built container images without any change.
+
+## Rejected alternatives
+
+There are several ways how to build container images.
+In addition to Kaniko and OpenShift Builds, I also looked at Buildah.
+But Kaniko and OpenShift Builds worked better (read as: _I didn't got Buildah to work_).
```

Comment:
I think what's important is that our API isn't bound to any particular tool, so we could always switch to something else at a later date.

Reply:
The suggested user API for the `KafkaConnect` is in the proposal. The internal code doing the builds will always depend on the exact tool being used. But it currently just builds a `Dockerfile` generated by the operator. So in theory any tool which can do that should do and it should be possible to implement it.